File: 0001-two-sum/0001-two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # track the numer required,
        # and the prev index
        d = {}

        for i in range(len(nums)):
            n = nums[i]
            # if n was required, we're done
            if n in d:
                return [i, d[n]]

            # else store the number as required
            # which will make target when added to n
            d[target-n] = i

        return []
        

File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    // let's have a map of { numberNeeded: prevIndex }
    let m = {}

    for(let i=0; i<nums.length; ++i) {
      const n = nums[i] // this number
      // if n is already required
      // we are done
      if(n in m) return [i, m[n]]

      // store its complimentary req num
      m[target-n] = i
    }

    return []
};